### PR TITLE
(QTM-190) - DropdownLight Implemented selectedItem, onChange

### PR DESCRIPTION
**Pull request by douglasjr97**
## Description
https://jirasoftware.catho.com.br/browse/QTM-190

## Review guide
- [ ] Unit tests (yarn test:components)
- [ ] Regression
- [ ] Code review

### CODE REVIEW

**Review by MarcosViniciusPC — CHANGES_REQUESTED** [review]
[] Code review
[] Unit

I imagine that it is necessary to include the ignoreSpecialChars and autocomplete properties for situations in which the dropdown has an extensive list of items. In this case, enabling the search greatly facilitates its use.

File: components/DropdownLight/DropdownLight.jsx [hunk 1]
```diff
@@ -241,8 +250,9 @@ const DropdownLight = ({ disabled, items, theme, placeholder, name }) => {
         type="text"
         hidden
         name={name}
-        defaultValue={selectedItem}
+        value={selectedOptionItem.value || selectedOptionItem}
         aria-label="selecione uma opção"
+        onChange={onChange}
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.jsx:255]: Is this attribute required?

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 2]
```diff
@@ -120,4 +120,19 @@ describe('<DropdownLight />', () => {
...
+    render(<DropdownLight items={itemsStringMock} onChange={onChangeMock} />);
+
+    userEvent.tab();
+    userEvent.keyboard(ENTER_KEY_CODE);
+
+    userEvent.keyboard(
+      `${ARROW_DOWN_KEY_CODE}${ARROW_DOWN_KEY_CODE}${ENTER_KEY_CODE}`,
+    );
+
+    expect(onChangeMock).toHaveBeenCalled();
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.unit.test.jsx:135]: Is this test necessary? Apparently the test below already validates the expected behavior

File: components/DropdownLight/index.d.ts [hunk 3]
```diff
@@ -14,6 +14,8 @@ export interface DropdownLightProps {
     spacing?: {};
     baseFontSize?: number;
   };
+  onChange?: (value: string) => void;
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/index.d.ts:17]: Is the onChange argument always a string?? Check this as I believe this is wrong

File: stories/DropdownLight/DropdownLight.stories.jsx [hunk 4]
```diff
@@ -17,3 +18,22 @@ export const Disabled = Template.bind({});
...
+    { value: 'Strawberry', label: 'Strawberry' },
+  ];
+  const [selectedFruit, setSelectedFruit] = useState({ label: 'Banana' });
+
+  return (
+    <DropdownLight
+      label="Fruits"
+      placeholder="Select a fruit"
+      items={fruits}
+      selectedItem={selectedFruit}
```
Comment (MarcosViniciusPC) [inline stories/DropdownLight/DropdownLight.stories.jsx:35]: selected item must be an object exactly the same as an item in the list, that is, it must be a string or an object with value and label properties. The object used here has only the label

File: stories/DropdownLight/DropdownLight.stories.jsx [hunk 5]
```diff
@@ -17,3 +18,22 @@ export const Disabled = Template.bind({});
 Disabled.args = {
   disabled: true,
 };
+
+export const Controlled = () => {
```
Comment (MarcosViniciusPC) [inline stories/DropdownLight/DropdownLight.stories.jsx:22]: good story

**Review by MarcosViniciusPC — CHANGES_REQUESTED** [review]
[] Code review
[] Unit
[x] Regression
[] Documentation

Include two tests, as requested in the previous review. The first is "must call onChange when an item in the list is selected" (in this case it is to select an item without using the keyboard) and another test for when the "selectedItem" property is used.
Also, I want to know your opinion about the general observation I made in the previous review about implementing ignoreSpecialChars

File: components/DropdownLight/index.d.ts [hunk 6]
```diff
@@ -14,6 +14,8 @@ export interface DropdownLightProps {
...
+  selectedItem?: ItemPropType;
```
Comment (rapahaeru) [inline components/DropdownLight/index.d.ts:18]: wheres the `name` prop?

File: components/DropdownLight/DropdownLight.unit.test.jsx (hunk 2 above)
Comment (douglasjr97) [reply components/DropdownLight/DropdownLight.unit.test.jsx:135]: I think it's necessary, I made an adaptation using the objects to ensure that onChange is working correctly.

```suggestion
    it('should call onChange when a list item is selected using keyboard', () => {
    const onChangeMock = jest.fn();
    render(<DropdownLight items={itemsObjectMock} onChange={onChangeMock} />);

    userEvent.tab();
    userEvent.keyboard(ENTER_KEY_CODE);

    userEvent.keyboard(
      `${ARROW_DOWN_KEY_CODE}${ARROW_DOWN_KEY_CODE}${ENTER_KEY_CODE}`,
    );

    expect(onChangeMock).toHaveBeenCalled();
    expect(onChangeMock).toHaveBeenCalledWith({label: 'Strawberry', value: 'Strawberry',});
  });
```

File: components/DropdownLight/DropdownLight.jsx [hunk 7]
```diff
@@ -241,8 +250,9 @@ const DropdownLight = ({ disabled, items, theme, placeholder, name }) => {
         type="text"
         hidden
         name={name}
-        defaultValue={selectedItem}
+        value={selectedOptionItem?.value || selectedOptionItem}
         aria-label="selecione uma opção"
+        onChange={onChange}
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.jsx:255]:
```suggestion
```
unnecessary

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 8]
```diff
@@ -120,4 +120,22 @@ describe('<DropdownLight />', () => {
...
+    render(<DropdownLight items={itemsObjectMock} onChange={onChangeMock} />);
+
+    userEvent.tab();
+    userEvent.keyboard(ENTER_KEY_CODE);
+
+    userEvent.keyboard(
+      `${ARROW_DOWN_KEY_CODE}${ARROW_DOWN_KEY_CODE}${ENTER_KEY_CODE}`,
+    );
+
+    expect(onChangeMock).toHaveBeenCalled();
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.unit.test.jsx:135]:
```suggestion
```
Redundant

File: components/DropdownLight/DropdownLight.unit.test.jsx [hunk 9]
```diff
@@ -120,4 +120,22 @@ describe('<DropdownLight />', () => {
...
+    expect(onChangeMock).toHaveBeenCalledWith({
+      label: 'Strawberry',
+      value: 'Strawberry',
+    });
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/DropdownLight.unit.test.jsx:139]: Good

File: components/DropdownLight/index.d.ts [hunk 10]
```diff
@@ -14,6 +14,10 @@ export interface DropdownLightProps {
     spacing?: {};
     baseFontSize?: number;
   };
+  selectedItem?: ItemPropType;
+  onChange?: (
+    selectedItem: ItemPropType | null,
```
Comment (MarcosViniciusPC) [inline components/DropdownLight/index.d.ts:19]:
```suggestion
    selectedItem?: ItemPropType,
```

**Review by MarcosViniciusPC — APPROVED** [review]
[x] Code review
[x] Unit
[x] Regression
[x] Documentation

File: stories/DropdownLight/DropdownLight.stories.jsx [hunk 11]
```diff
@@ -17,3 +18,27 @@ export const Disabled = Template.bind({});
...
+  const fruits = [
```
Comment (rapahaeru) [inline stories/DropdownLight/DropdownLight.stories.jsx:23]: dont we have the same object in mock file? How about of use it here?

File: components/DropdownLight/index.d.ts (hunk 6 above)
Comment (rapahaeru) [reply components/DropdownLight/index.d.ts:17]: :warning:

**Review by rapahaeru — CHANGES_REQUESTED** [review]
## Review guide
- [x] Unit tests (yarn test:components)
- [x] Regression
- [ ] Code review

The component shows this warning in console. I think it should be fixed.
image

tests doesnt pass here:
image
Its seems unnecessary code, how about this:
```
    if (document.activeElement !== buttonRef.current && enterPress && listOptions.current) {
      console.log(':: ', listOptions.current)
      setIsOpen(false);

      const itemsList = [...listOptions.current.children];

      if (itemsList.some(element => element === document.activeElement)) {
        selectItem(items[cursor]);
      }
    }
```

File: stories/DropdownLight/DropdownLight.story.mdx [hunk 12]
```diff
@@ -48,3 +49,9 @@ import { DropdownLight } from '@catho/quantum';
 <Canvas>
   <Story story={WithArrayOfString}/>
 </Canvas>
+
+### Controlled
```
[the remark this answers is not in the document]
Comment (douglasjr97) [reply stories/DropdownLight/DropdownLight.story.mdx:53]: i agree

**Review by rapahaeru — CHANGES_REQUESTED** [review]
Is suggested codes codes thats was not implemented.

**Review by rapahaeru — CHANGES_REQUESTED** [review]
There are unfixes suggestions.
please, read the message with the code suggestion again.

**Review by rapahaeru — APPROVED** [review]

**Pull request merged**